GitlabCI

Durf1985 opened the pull request:
## In the process done:

* Created an image with `docker` and `docker-compose` preinstalled
* An infrastructure has been created for deploying `gitlab-omnibus` using `terraform` using `storage-bucket` and `backend`
* Created playbooks and roles for installing `gitlab-omnibus`, `gitlab-runner` 
* A playbook has been created to mount a disk to a VM running `gitlab-omnibus` and `gitlab-runner`
* A playbook has been created to launch `reddit-monolith` using `docker-compose`
* A pipeline has been created in which images used to launch reddit-monolith are created and tested using terraform and GCP

## How to setup the project:

To start the pipeline, you need to run `git push gitlab-ssh gitlab-ci-1` in your gitlab-ci omnibus 

## How to check the project:

* To check gitlab-ci  follow the link <http://your_VM_for_gitlab_ci_external_ip>
* After starting the pipeline or apply the roles of reddit_microservice => follow the link <http://your_VM_for_reddit_app_external_ip:9292> in order to test your application

## PR checklist
The label with the homework number, in label description - homework lecture and topic

Durf1985 committed:
DO-126-128 Gitlab ci

* Added a playbook and a role for baking an image with docker_compose installed
* Added a role and a playbook for installing and running the Gitlab
	container using docker_compose
* Added a test for the docker_compose role
* The terraform main configuration file has been changed.tf and variable
	file variables.tf
* Changed packer template (added docker_compose installation)
* Corrections of ansible files have been made in accordance with the
	recommendations of ansible-lint

Durf1985 committed:
Add reddit app

Durf1985 committed:
DO-128/129 Setup a basic CI

Durf1985 committed:
DO-130 Setup dynamic environments

Durf1985 committed:
Demo commit

loktionovam commented on docker-monolith/infra/ansible/roles/gitlab_omnibus/defaults/main.yml:
```diff
@@ -0,0 +1,3 @@
+---
+env: local
+gitlab_ci_external_ip: 0.0.0.0
```
You renamed the role but forget to rename variable prefixes. There (and other places) should be `gitlab_omnibus_external_ip`

loktionovam commented on docker-monolith/infra/ansible/roles/gitlab_omnibus/templates/docker-compose.yml.j2:
```diff
@@ -0,0 +1,17 @@
+version: '3.6'
+services:
+  web:
+    image: 'gitlab/gitlab-ce:latest'
```
it's better not to use latest tag here and pin to specific version

Durf1985 committed:
DO-130/131 Setup dynamic environments/per branch
Simplified gitlab-ci.yml to make it clear to me what works
and how it works. The rest of this file is stored
in a separate repository outside the repository. It will be added
as soon as it is ready
Known Issue: Variable assignment changed in .env.example file
to hardcode, having problems with gitlab variables when migrating with ansible to
VM on which the application will be deployed

Durf1985 committed:
DO-130/131 Setup dynamic environments/per branch
Simplified gitlab-ci.yml to make it clear to me what works
and how it works. The rest of this file is stored
in a separate repository outside the repository. It will be added
as soon as it is ready
Known Issue: Variable assignment changed in .env.example file
to hardcode, having problems with gitlab variables when migrating with ansible to
VM on which the application will be deployed

loktionovam commented on src/docker-compose copy.yml:
```diff
@@ -0,0 +1,42 @@
+version: '3.3'
```
What is this file?

loktionovam commented on src/docker-compose.yml:
```diff
@@ -1,16 +1,16 @@
     image: ${DOCKER_HUB_USERNAME}/ui:${UI_VERSION}
     ports:
-      - ${SRC_PORT}:${DEST_PORT}/tcp
+      - 9292:9292
```
Why you use hardcode here again?

loktionovam commented on docker-monolith/infra/ansible/playbooks/attach_disk.yml:
```diff
@@ -0,0 +1,25 @@
+---
+- name: Launcher for container on VM instaces
```
Name of this playbook should be like "Format and mount GCP disk" rather than "Launcher for container on VM instances"

loktionovam commented on docker-monolith/infra/ansible/roles/gitlab_runner/tasks/main.yml:
```diff
@@ -0,0 +1,35 @@
+  ansible.builtin.command:
+    "docker run -d --name gitlab-runner \
+    --restart always \
+    -v /srv/gitlab-runner/config:/etc/gitlab-runner \
```
Where the path `/srv/gitlab-runner` is created?

loktionovam commented on docker-monolith/infra/ansible/roles/gitlab_runner/tasks/main.yml:
```diff
@@ -0,0 +1,35 @@
+    --restart always \
+    -v /srv/gitlab-runner/config:/etc/gitlab-runner \
+    -v /var/run/docker.sock:/var/run/docker.sock \
+    gitlab/gitlab-runner:latest"
```
Create ansible variable and pin runner version rather than using latest

loktionovam commented on docker-monolith/infra/ansible/roles/gitlab_runner/tasks/main.yml:
```diff
@@ -0,0 +1,35 @@
+    --run-untagged --locked=false \
+    --docker-privileged \
+    --non-interactive \
+    --url http://34.136.57.185/ \
```
Remove hardcode and create variable for url

loktionovam commented on docker-monolith/infra/ansible/roles/gitlab_runner/tasks/main.yml:
```diff
@@ -0,0 +1,35 @@
+    --non-interactive \
+    --url http://34.136.57.185/ \
+    --registration-token {{gitlab_runner_registration_token}} \
+    --description my-runner
```
Parametrize the description too

Durf1985 committed:
DO-130/131 Setup dynamic environments/per branch

loktionovam commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,116 @@
+    - echo "$APPUSER" > ${CI_PROJECT_DIR}/appuser
+    - chmod 0600 ${CI_PROJECT_DIR}/appuser
+    - sed -i '\#vault_password_file = ~/.ansible/vault.key#d' $ANSIBLE_CONFIG
+    - sed -i 's|private_key_file = ~/.ssh/appuser|private_key_file = '${CI_PROJECT_DIR}/appuser'|g' $ANSIBLE_CONFIG
```
FYI: You can pass path to private key (and many more) using ansible-playbook parameters like:

`ansible-playbook --private-key ~/.ssh/appuser|private_key_file`

And these parameters have higher precedence than those in ansible.cfg

loktionovam commented on docker-monolith/infra/ansible/playbooks/omnibus_password:
```diff
@@ -0,0 +1,18 @@
+---
```
Playbooks should have .yml extension

loktionovam commented on docker-monolith/infra/ansible/playbooks/reddit_microservice.yml:
```diff
@@ -0,0 +1,17 @@
+        timeout: 300
+
+
+- name: Wait for infrastructure to become available
```
Here should be "Deploy reddit microservices" (or similar to) because you are waiting infrastructure readiness above.

loktionovam commented on docker-monolith/infra/ansible/roles/reddit_microservice/defaults/main.yml:
```diff
@@ -0,0 +1,6 @@
+env: local
+deploy_user: "foobar"
+reddit_microservice_vm_ip: 0.0.0.0
+reddit_microservice_docker_image_tag: "foobar"
```
Place here well-known `latest`

Durf1985 committed:
Added the README file
